frontend/javascripts/viewer/model/sagas/meshes/ad_hoc_mesh_saga.ts

```diff
@@ -0,0 +1,673 @@
+  const tracingStoreHost = yield* select((state) => state.annotation.tracingStore.url);
+  const dataStoreUrl = `${dataStoreHost}/data/datasets/${owningOrganization}/${datasetDirectoryName}/layers/${
+    layer.fallbackLayer != null ? layer.fallbackLayer : layer.name
+  }`;
+  const tracingStoreUrl = `${tracingStoreHost}/tracings/volume/${layer.name}`;
+
+  const mag = magInfo.getMagByIndexOrThrow(zoomStep);
+
+  if (isInitialRequest) {
+    sendAnalyticsEvent("request_isosurface", {
+      mode: useDataStore ? "view" : "annotation",
+    });
+  }
+
+  let retryCount = 0;
+
+  const { segmentMeshController } = getSceneController();
+
+  const cubeSize = marchingCubeSizeInTargetMag();
+
+  while (retryCount < MAX_RETRY_COUNT) {
+    try {
+      const { buffer: responseBuffer, neighbors } = yield* call(
+        {
+          context: null,
+          fn: computeAdHocMesh,
+        },
+        useDataStore ? dataStoreUrl : tracingStoreUrl,
+        {
+          position: clippedPosition,
+          additionalCoordinates: additionalCoordinates || undefined,
+          mag,
+          segmentId,
+          cubeSize,
+          scaleFactor,
+          findNeighbors,
+          ...meshExtraInfo,
+        },
+      );
+      const vertices = new Float32Array(responseBuffer);
+
+      if (removeExistingMesh) {
+        segmentMeshController.removeMeshById(segmentId, layer.name);
+      }
+
+      const opacity = meshExtraInfo.opacity || Constants.DEFAULT_MESH_OPACITY;
+
+      // We await addMeshFromVerticesAsync here, because the mesh saga will remove
+      // an ad-hoc loaded mesh immediately if it was "empty". Since the check is
+      // done by looking at the scene, we await the population of the scene.
+      // Theoretically, this could be built differently so that other ad-hoc chunks
+      // can be loaded in parallel to addMeshFromVerticesAsync. However, it's unclear
+      // how big the bottleneck really is.
+      yield* call(
+        { fn: segmentMeshController.addMeshFromVerticesAsync, context: segmentMeshController },
+        vertices,
+        segmentId,
+        layer.name,
+        opacity,
+        additionalCoordinates,
+      );
+      return neighbors.map((neighbor) =>
+        getNeighborPosition(clippedPosition, neighbor, zoomStep, magInfo),
+      );
+    } catch (exception) {
+      retryCount++;
+      ErrorHandling.notify(exception as Error);
+      console.warn("Retrying mesh generation due to", exception);
+      yield* call(sleep, RETRY_WAIT_TIME * 2 ** retryCount);
+    }
+  }
+
+  return [];
+}
+
+function* markEditedCellAsDirty(): Saga<void> {
+  const volumeTracing = yield* select((state) => getActiveSegmentationTracing(state));
+
+  if (volumeTracing != null && volumeTracing.fallbackLayer == null) {
+    const activeCellId = volumeTracing.activeCellId;
+    modifiedCells.add(activeCellId);
+  }
+}
+
+function* refreshMeshes(): Saga<void> {
```

Comment:
this function is not called, because in the segment view, separate `refreshMeshAction`s are dispatched. @philippotto do you know if there is a reason for that? should I change this?

Reply:
it looks like this function is only called when the front-end `api.data.refreshMeshes` is used. you could add a comment to the function saying exactly this.
i don't think you need to change this. the segments view probably uses the singular `refreshMeshAction` because the user typically only refreshes individual meshes. for the case, where the user wants to refresh all meshes, this function could be used but I don't see a big downside, either.